docs/proposals-accepted/20240909-live-blocks-in-objectstorage.md

```diff
@@ -0,0 +1,60 @@
+---
+type: proposal
+title: Avoid querying blocks in object storage if they have a live source
+status: approved
+owner: MichaHoffmann
+menu: proposals-accepted
+---
+
+## Avoid querying blocks in object storage if they have a live source
+
+> TL;DR: We propose to not query blocks in object storage that are still served by a Sidecar/Ruler or Receiver.
+
+## Why
+
+Accessing blocks from object storage is generally orders of magnitude slower then accessing the same data from other sources that can query them from disk or from memory. If the source that originally uploaded the blocks is still alive and still has access to the block then we would throw away the expensively obtained data from object storage during deduplication. Note that this will not put additional pressure on the source components since they would get queried during fan-out anyway. As an example, imagine a Sidecar to a Prometheus server that has 3 months of retention and a Storage Gateway that is configured to serve the whole range. Right now we dont have a great way to deal with this dynamically. This proposal should address this hopefully.
```

Comment:
There is somewhat of a way to deal with this through `--min-time` and `--max-time` flags but it is not ideal:

- What if some Sidecar/Receiver fails to upload blocks? Receiver doesn't delete blocks until they are uploaded so it might accidentally fall out of the configured min/max time.
- It almost requires all components (Receive/Sidecar/Ruler) to have equal retention to make `--max-time` work properly; otherwise, you might have to have multiple Thanos Store replicas with different `--max-time` and selector labels.

We would like to improve this user experience. I would add this to the Why section.